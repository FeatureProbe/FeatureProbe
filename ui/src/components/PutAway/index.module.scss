.put-away {
  position: absolute;
  bottom: 10px;
  font-size: 13px;
  color: #adb5bc;
  line-height: 32px;
  width: 180px;
  cursor: pointer;
  padding: 0 20px;
}

.put-away-close {
  width: 80px;
  text-align: center;
  padding: 0;
}

.put-away-divider {
  width: 160px;
  height: 1px;
  opacity: 0.2;
  background-image: linear-gradient(180deg, #74788d 0%, #F4F4F8 100%);
}

.put-away-divider-close {
  width: 60px;
  position: absolute;
  left: 10px;
}

.put-away-text {
  vertical-align: middle;
  margin-left: 8px;
}

.iconfont {
  font-size: 18px !important;
  vertical-align: middle;
}

.menu {
  vertical-align: middle;
  margin-left: 8px;
}