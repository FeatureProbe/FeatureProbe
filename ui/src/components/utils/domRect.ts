export class DOMRect {
    x: number;
    width: number;
    y: number;
    height: number;
    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}