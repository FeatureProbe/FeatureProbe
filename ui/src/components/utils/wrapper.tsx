import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
