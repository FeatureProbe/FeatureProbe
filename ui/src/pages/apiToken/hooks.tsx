import { useState } from 'react';
import { InputOnChangeData, TextAreaProps, DropdownProps, CheckboxProps } from 'semantic-ui-react';
import { useForm } from 'react-hook-form';
import { IToken } from 'interfaces/token';

export const useTokenInfo = () => {
  const [tokenInfo, saveTokenInfo] = useState<IToken>({
    name: '',
    role: undefined,
  });

  const [ originTokenInfo, saveOriginTokenInfo ] = useState<IToken>({
    name: '',
    role: undefined
  });

  const handleChange = (detail: InputOnChangeData | TextAreaProps | DropdownProps | CheckboxProps, type: string) => {
    tokenInfo[type] = detail.value as string;

    saveTokenInfo({...tokenInfo});
  };

  const init = () => {
    saveTokenInfo({
      name: '',
      role: undefined
    });
    saveOriginTokenInfo({
      name: '',
      role: undefined
    });
  };

  return {
    tokenInfo,
    originTokenInfo,
    init,
    handleChange,
    saveTokenInfo,
    saveOriginTokenInfo,
  };
};

export const useReactHookForm = () => {
  return {
    ...useForm(),
  };
};
