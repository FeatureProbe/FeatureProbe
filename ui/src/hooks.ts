import { useCallback, useEffect, useRef } from 'react';
import { FieldErrors } from 'react-hook-form';
import { createContainer } from 'unstated-next';
import { useLocalStorage } from 'utils/hooks';

export const useI18N = () => {
  const localLanguage = navigator.language.toLocaleLowerCase().slice(0, 2);
  const defaultLanguage = localLanguage === 'zh' ? 'zh-CN' : 'en-US';
  const [i18n, setI18n] = useLocalStorage('i18n', defaultLanguage);

  return {
    i18n,
    setI18n,
  };
};

export const I18NContainer = createContainer(useI18N);

export const useRequestTimeCheck = () => {
  const requestTimes = useRef(new Map());

  const creatRequestTimeCheck = useCallback((key: string) => {
    const requestTime = Date.now();
    requestTimes.current.set(key, requestTime);

    const check = () => {
      return requestTimes.current.get(key) === requestTime;
    };

    return check;
  }, [requestTimes]);

  return creatRequestTimeCheck;
};

export const useFormErrorScrollIntoView = (errors?: FieldErrors) => {
  const errorRef = useRef(errors);
  const beforeScroll = useRef<(names: string[]) => void>();

  useEffect(() => {
    errorRef.current = errors;
  }, [errors]);

  const scrollToError = useCallback(() => {
    if(errorRef.current) {
      const names = Object.keys(errorRef.current);
      if(beforeScroll.current) {
        beforeScroll.current(names);
      }
      setTimeout(() => {
        document.querySelector(`[name=${names[0]}]`)?.scrollIntoView({ behavior: 'smooth', block: 'center' });
      });
    }
  }, [errorRef]);

  const registerErrorName = useCallback((name: string) => {
    return {
      ref: (ref: unknown) => {
        (ref as HTMLElement)?.setAttribute('name', name);
      }
    };
  }, []);

  const setBeforeScrollCallback = useCallback((fc: (names: string[]) => void) => {
    beforeScroll.current = fc;
  }, [beforeScroll]);

  return {
    scrollToError,
    registerErrorName,
    setBeforeScrollCallback,
  };
};