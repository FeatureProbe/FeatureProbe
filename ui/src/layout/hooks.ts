import { useState } from 'react';
import { createContainer } from 'unstated-next';
import { useLocalStorage } from 'utils/hooks';

export const usePutaway = () => {
  const [ isPutAway, setIsputAway ] = useLocalStorage('sidebarIsPutAway', false);

  return { 
    isPutAway, 
    setIsputAway, 
  };
};

export const useUserInfo = () => {
  const [ userInfo, saveUserInfo ] = useState({
    account: '',
    role: '',
  });

  const [ approvalCount, saveApprovalCount ] = useState<number>(0);

  return {
    userInfo,
    approvalCount,
    saveUserInfo,
    saveApprovalCount,
  };
};

export const SidebarContainer = createContainer(usePutaway);
export const HeaderContainer = createContainer(useUserInfo);
