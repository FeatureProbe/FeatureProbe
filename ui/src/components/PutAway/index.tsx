import { useEffect } from 'react';
import classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import Icon from 'components/Icon';
import { SidebarContainer } from 'layout/hooks';
import styles from './index.module.scss';

const PutAway = () => {
  const {
    isPutAway,
    setIsputAway
  } = SidebarContainer.useContainer();

  const putAwayCls = classNames(
    styles['put-away'],
    {
      [styles['put-away-close']]: isPutAway
    }
  );

  const putAwayDividerCls = classNames(
    styles['put-away-divider'],
    {
      [styles['put-away-divider-close']]: isPutAway
    }
  );

  useEffect(() => {
    setIsputAway(isPutAway);
    // TODO: better way
    const footerDom = document.getElementById('footer');
    if (footerDom) {
      if (isPutAway) {
        footerDom.style.left = '100px';
      } else {
        footerDom.style.left = '220px';
      }
    }
  }, [isPutAway, setIsputAway]);

  return (
    <div className={putAwayCls} onClick={() => {setIsputAway(!isPutAway);}}>
      <div className={putAwayDividerCls}></div>
      {
        isPutAway 
          ? <Icon customclass={styles.iconfont} type='put-up' /> 
          : <Icon customclass={styles.iconfont} type='put-away' />
      }
      {
        !isPutAway && <span className={styles.menu}>
          <FormattedMessage id='sidebar.putaway.text' />
        </span>
      }
    </div>
	);
};

export default PutAway;