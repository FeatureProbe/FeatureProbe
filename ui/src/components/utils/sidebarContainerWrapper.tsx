import { SidebarContainer } from 'layout/hooks';

export default SidebarContainer.Provider;
