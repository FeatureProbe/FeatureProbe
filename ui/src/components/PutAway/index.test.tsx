import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { SidebarContainer } from 'layout/hooks';
import { useEffect } from 'react';
import { IntlWrapper } from 'components/utils/wrapper';
import PutAway from '.';

const TrueWrapper: React.FC = (props) => {
  return (
    <IntlWrapper>
      <SidebarContainer.Provider>
        <UseHookWrapper isPutAway={true}>
          {props.children}
          <div id="footer"></div>
        </UseHookWrapper>
      </SidebarContainer.Provider>
    </IntlWrapper>
  );
};

const FalseWrapper: React.FC = (props) => {
  return (
    <IntlWrapper>
      <SidebarContainer.Provider>
        <UseHookWrapper isPutAway={false}>{props.children}</UseHookWrapper>
      </SidebarContainer.Provider>
    </IntlWrapper>
  );
};

const UseHookWrapper: React.FC<{ isPutAway?: boolean }> = ({ children, isPutAway }) => {
  const { setIsputAway } = SidebarContainer.useContainer();

  useEffect(() => {
    setIsputAway(isPutAway);
  }, [isPutAway]);

  return <>{children}</>;
};

it('PutAway true snapshot', (done) => {
  (async () => {
    const { baseElement } = render(<PutAway />, {
      wrapper: TrueWrapper,
    });
    expect(baseElement).toMatchSnapshot();

    done();
  })();
});

it('PutAway false snapshot', (done) => {
  (async () => {
    const { baseElement } = render(<PutAway />, {
      wrapper: FalseWrapper,
    });
    expect(baseElement).toMatchSnapshot();

    done();
  })();
});

test('click PutAway', (done) => {
  (async () => {
    const { asFragment } = render(<PutAway />, {
      wrapper: FalseWrapper,
    });

    await userEvent.click(screen.getByText('Collapse'));
    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});
