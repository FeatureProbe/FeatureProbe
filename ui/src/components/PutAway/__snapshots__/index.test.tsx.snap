// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PutAway false snapshot 1`] = `
<body>
  <div>
    <div
      class="put-away"
    >
      <div
        class="put-away-divider"
      />
      <span
        class="iconfont icon-put-away iconfont"
        customclass="iconfont"
        type="put-away"
      />
      <span
        class="menu"
      >
        Collapse
      </span>
    </div>
  </div>
</body>
`;

exports[`PutAway true snapshot 1`] = `
<body>
  <div>
    <div
      class="put-away put-away-close"
    >
      <div
        class="put-away-divider put-away-divider-close"
      />
      <span
        class="iconfont icon-put-up iconfont"
        customclass="iconfont"
        type="put-up"
      />
    </div>
    <div
      id="footer"
      style="left: 100px;"
    />
  </div>
</body>
`;

exports[`click PutAway 1`] = `
<DocumentFragment>
  <div
    class="put-away put-away-close"
  >
    <div
      class="put-away-divider put-away-divider-close"
    />
    <span
      class="iconfont icon-put-up iconfont"
      customclass="iconfont"
      type="put-up"
    />
  </div>
</DocumentFragment>
`;
